### DISCUSSION TITLE
About multiplayer and game controller (joystick)

### DISCUSSION THREAD

**Discussion by alberic89** [SuperTux/supertux#3020] [General]
Sorry for the bad English, it is not my language.

I tried the new feature that are multiplayer game and controller support, and I have many commentary to do about these. (I know that is totally work-in-progress and alpha features, so it is normal if all is not perfect. But it is surprisingly good.)

First, the game controller.
I tried to use a Wiimote to play to Supertux, because it is a cool controller and the only one I have. So, I connected my Wiimote to my computer with Bluetooth, and I open the game.
But, instead of one controller, I see three : the good one, the IR camera of the Wiimote, and the accelerometers. The game thought I connected three controllers instead of one.
In the state, this is not grave, but it can have unattended side effects, as we will see after.
After choosing the right controller, the game is totally playable and very funny. 🐧

Now, the multiplayer.
Adding multiplayer support is a great idea, playing with someone else to Supertux is _very_ fun (especially when there are throwaway objects and badguys in the level 😈 ).
The camera is greatly mastered when there are multiple players, and it is a very positive point 👍🏻 (even if I prefer something like in Kirby, not as in Mario Bros).
But, in some case, multiplayer looks not finished. Especially in the cutscenes, if there are two players, the second player can move as he wants during the cutscene. One solution to this problem can be to despawn all players except one, hide the indicator, block all player input, and let the others players to respawn with their power-up after the cutscene.
Another black point is the keyboard: if multiple player use the same keyboard, there is a limited number of key that the players can press. Pressing more keys with result in nothing, and it blocked me during advanced maneuver, such a carry a block, roll, and jump in the same time as the other player.

Finally, if I want to use Wiimote with multiplayer, what will append ? It will resolve the problem with the keyboard, no ?
In theory.
But there is another issue. As we said, the game think I connected three controllers for one Wiimote. So, if I connect two Wiimotes, there are six controllers in the game. Then, the game said “Ho, there are six controllers for this multiplayer game, so let spawn six players !”. I tried to remove players and select the right controllers, but it never worked because I need to restart the game after choosing a controller for a player, and the number of players is re-computed at each game start. 😞

So, that is the end of my personal story. Thanks to all contributors for their incredible works !

**Comment by MatusGuy** [comment]
> But, instead of one controller, I see three

Not sure if we're planning to properly support Wiimotes. I'm curious, what operating system are you using?

> Pressing more keys with result in nothing

I think some keyboards limit the amount of keys pressed at the same time.

**Reply by alberic89** [reply]
> Not sure if we're planning to properly support Wiimotes. I'm curious, what operating system are you using?

I am using Linux (Debian 12), and it surprised me because in my head, SDL properly manage Wiimotes. If my memory is good, SupertuxKart fully support Wiimote. And I used a Bluetooth USB adaptor to connect the Wiimotes to my PC, a Realtek one.

But I understood that you can't support all existing controller, especially old one. If I want my controller in the game, I need to make it possible by myself. Maybe one day, when I have more time and C/C++ skills…

**Reply by MatusGuy** [reply]
SDL only supports generic joypads or game controllers. No wiimotes or other exotic controllers like that. SuperTuxKart has custom wiimote support.
